src/bun.js/test/expect.zig

```diff
@@ -3948,10 +4459,382 @@ pub const ExpectArrayContaining = struct {
     }
 };
 
+/// An instantiated asymmetric custom matcher, returned from calls to `expect.toCustomMatch(...)`
+///
+/// Reference: `AsymmetricMatcher` in https://github.com/jestjs/jest/blob/main/packages/expect/src/types.ts
+/// (but only created for *custom* matchers, as built-ins have their own classes)
+pub const ExpectCustomAsymmetricMatcher = struct {
+    pub usingnamespace JSC.Codegen.JSExpectCustomAsymmetricMatcher;
+
+    flags: Expect.Flags = .{},
+
+    pub fn finalize(
+        this: *ExpectCustomAsymmetricMatcher,
+    ) callconv(.C) void {
+        VirtualMachine.get().allocator.destroy(this);
+    }
+
+    /// Implements the static call of the custom matcher (`expect.myCustomMatcher(<args>)`),
+    /// which creates an asymmetric matcher instance (`ExpectCustomAsymmetricMatcher`).
+    /// This will not run the matcher, but just capture the args etc.
+    pub fn create(globalObject: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame) callconv(.C) JSValue {
+        defer globalObject.bunVM().autoGarbageCollect();
+
+        var flags: Expect.Flags = undefined;
+
+        // try to retrieve the ExpectStatic instance (to get the flags)
+        var thisValue: JSValue = callFrame.this();
+        if (ExpectStatic.fromJS(thisValue)) |expect_static| {
+            flags = expect_static.flags;
+        } else {
+            // if it's not an ExpectStatic instance, it was called from the Expect constructor, so assume default flags
+            flags = .{};
+        }
+
+        // retrieve the user-provided matcher function
+        var func: JSValue = callFrame.callee();
+        var matcher_fn = func.get(globalObject, Expect.custom_matcher_fn_property) orelse unreachable;
```

Comment:
We can use a `JSC::JSBoundFunction` here I think. JSBoundFunction can preserve the `this` and any additional arguments. `JSBoundFunction` gets inlined in the JIT, so it has very little overhead.

If we cannot use a JSBoundFunction, then the backup would be to store the custom matcher as a `WriteBarrier` on the instance itself. Then we can avoid the (observable) property access.

Reply:
I tried with a bound function but got no way to recover the Expect/ExpectStatic instance to retrieve the flags as you need to specify the `this` on creation (i.e. in the expect.extends execution).

I will do the second way, I thought about subclassing `JSFunction` but didn't know how to go about it, thanks for the tip!

Reply:
Another option would be to make custom matchers a `CustomGetterSetter` which returns the JSFunction subclass.

It would work like [EventListenerMap in WebCore](https://github.com/oven-sh/WebKit/blob/0aa1f6dfc9b48be1e89b8f709fd44e3c88feaf33/Source/WebCore/dom/EventListenerMap.h#L81-L89):


```c++
template<typename Visitor>
void EventListenerMap::visitJSEventListeners(Visitor& visitor)
{
    Locker locker { m_lock };
    for (auto& entry : m_entries) {
        for (auto& eventListener : entry.second)
            eventListener->callback().visitJSFunction(visitor);
    }
}
```

The functions would lazily be created on each instance on access. Then, cached and visited within [`visitAdditionalChildren`](https://github.com/oven-sh/WebKit/blob/0aa1f6dfc9b48be1e89b8f709fd44e3c88feaf33/Source/WebCore/bindings/js/JSWorkerGlobalScopeCustom.cpp#L40-L54) on each instance:
```c++
template<typename Visitor>
void JSWorkerGlobalScope::visitAdditionalChildren(Visitor& visitor)
{
    if (auto* location = wrapped().optionalLocation())
        addWebCoreOpaqueRoot(visitor, *location);
    if (auto* navigator = wrapped().optionalNavigator())
        addWebCoreOpaqueRoot(visitor, *navigator);
    ScriptExecutionContext& context = wrapped();
    addWebCoreOpaqueRoot(visitor, context);
    
    // Normally JSEventTargetCustom.cpp's JSEventTarget::visitAdditionalChildren() would call this. But
    // even though WorkerGlobalScope is an EventTarget, JSWorkerGlobalScope does not subclass
    // JSEventTarget, so we need to do this here.
    wrapped().visitJSEventListeners(visitor);
}
```

Reply:
Did it with the JSFunction subclassing approach, seems to work fine!
This way you can access both the the expect instance to get the flags (`callframe.this()`) and recover the user-provided matcher function from `calframe.callee()` which is the new subclass instance